# Processing the CoreLogic Data using PySpark

In this Jupyter notebook, we demonstrate how you can process the CoreLogic data using PySpark. 

In particular, we will show you how to:
* import the data, 
* explore rows and columns, 
* filter for a given location, 
* inspect the data for missing and erroneous information, 
* correct data errors,
* and write a subset of the data to a csv file for later use. 

This [PySpark cheat sheet](http://datacamp-community-prod.s3.amazonaws.com/acfa4325-1d43-4542-8ce4-bea2d287db10) provides a great overview of available PySpark functionality. 

The notebook was created using a 'Jupyter + Spark Basic session' in [Open on Demand](https://arc.umich.edu/open-ondemand/) (OOD) on [Great Lakes](https://arc.umich.edu/greatlakes/) (GL). This automatically initializes Spark in the background.

If you are not running the notebook using OOD on GL, you will most likely have to make sure PySpark is installed on your system and then initialize it by typing 

```from pyspark import SparkContext```

```sc = SparkContext(master = 'local[2]')```

If you encounter any errors or have questions about this Jupyter notebook, or if you would like us to add another PySpark example using the CoreLogic data, feel free to reach out to [Armand Burks](mailto:arburks@umich.edu) and [Jule Krüger](mailto:julianek@umich.edu).

## Importing the CoreLogic data with PySpark

The CoreLogic data are stored in a Turbo volume on the ```/nfs``` drive. To get access, you will need to sign a Memorandum of Understanding (MOU) with the University of Michigan Library. For more information about this, see [here](https://github.com/arc-ts/corelogic-on-greatlakes/tree/main/intro-to-corelogic-data). To execute this notebook, you will have to be granted access to the CoreLogic data on Turbo.

The raw data come in three separate files: deeds (28GB), foreclosures (6GB), and taxes (24GB). [CSCAR](https://cscar.research.umich.edu/) pre-processed the CoreLogic data and stored each raw file in 100 separate partitions. You can read more about this methodology in ```nfs/turbo/lib-data-corelogic/Docs/cscar_data.txt```.

We are going to work with the pre-processed data to improve import speeds. Let's store the paths to the pre-processed partitioned files in three separate variables:

In [1]:
turbo_path_fcl = "/nfs/turbo/lib-data-corelogic/Data/fcl/*.gz"
turbo_path_deed = "/nfs/turbo/lib-data-corelogic/Data/deed/*.gz"
turbo_path_tax = "/nfs/turbo/lib-data-corelogic/Data/tax/*.gz"

The following command imports a set of partitioned CoreLogic data files into a dataframe (df). You can switch between foreclosures, deeds, and taxes by choosing the relevant path accordingly. In each text file, the pipe ```|``` was used as a delimiter and is set  in the ```sep``` argument as such.

__A note on import speeds__: Because of their size, it takes a little while to read in each partitioned data file collection. Importing the raw data files takes much longer. Using the pre-processed data from CSCAR, which distributes the data across 100 partitions, greatly improves import speeds as Spark can spread computation across multiple cores.

Let's work with the deed data for now:

In [2]:
df = spark.read.csv(turbo_path_deed, header=True, sep="|")

(A small note on reading in the data: You could set the ```inferSchema=True``` argument, which would make Python infer the column type. The drawback of this is that FIPS and ZIP codes that start with a `0` have the leading `0` removed in the creation of an integer variable. We choose to keep ```inferSchema=False```, the default argument, to avoid this.)

## Exploring the CoreLogic data (rows and columns)

First, we want to get a basic idea of the CoreLogic data. What are the columns and what types of values do they contain? The .printSchema() method prints the data type (integer, string, double) for each column in the dataset. 

Because we set ```inferSchema=False```, all columns are read in as string data types. ```nullable = true``` means that a given column can accept missing (```null```) values.

*Note*: FIPS stands for "Federal Information Processing Series" codes that identify geographic entities within the U.S. ([source](https://www.census.gov/library/reference/code-lists/ansi.html)). We will be working with FIPS codes later in this notebook.

In [3]:
df.printSchema()

root
 |-- FIPS: string (nullable = true)
 |-- APN (Parcel Number) (unformatted): string (nullable = true)
 |-- PCL ID IRIS FORMATTED: string (nullable = true)
 |-- APN SEQUENCE NUMBER: string (nullable = true)
 |-- PENDING RECORD INDICATOR: string (nullable = true)
 |-- CORPORATE INDICATOR: string (nullable = true)
 |-- OWNER FULL NAME: string (nullable = true)
 |-- OWNER 1 LAST NAME: string (nullable = true)
 |-- OWNER 1 FIRST NAME & M I: string (nullable = true)
 |-- OWNER 2 LAST NAME: string (nullable = true)
 |-- OWNER 2 FIRST NAME & MI: string (nullable = true)
 |-- OWNER ETAL INDICATOR: string (nullable = true)
 |-- C/O NAME: string (nullable = true)
 |-- OWNER RELATIONSHIP RIGHTS CODE: string (nullable = true)
 |-- OWNER RELATIONSHIP TYPE: string (nullable = true)
 |-- PARTIAL INTEREST INDICATOR: string (nullable = true)
 |-- ABSENTEE OWNER STATUS: string (nullable = true)
 |-- PROPERTY LEVEL LATITUDE: string (nullable = true)
 |-- PROPERTY LEVEL LONGITUDE: string (nullable = tr

Let's calculate how many rows and columns there are in total. Note that we will store these values in their own variables because it takes quite a while to calculate the number of rows in our data. If we commit this number to current memory by storing it in a variable, we can easily reuse it later for various purposes without having to recalculate it first. This practice is also known as "lazy execution."

In [4]:
numRows = df.count()
numCols = len(df.columns)
print(numRows,numCols)

367782480 97


We can print a list of the column names like so.

In [5]:
df.columns

['FIPS',
 'APN (Parcel Number) (unformatted)',
 'PCL ID IRIS FORMATTED',
 'APN SEQUENCE NUMBER',
 'PENDING RECORD INDICATOR',
 'CORPORATE INDICATOR',
 'OWNER FULL NAME',
 'OWNER 1 LAST NAME',
 'OWNER 1 FIRST NAME & M I',
 'OWNER 2 LAST NAME',
 'OWNER 2 FIRST NAME & MI',
 'OWNER ETAL INDICATOR',
 'C/O NAME',
 'OWNER RELATIONSHIP RIGHTS CODE',
 'OWNER RELATIONSHIP TYPE',
 'PARTIAL INTEREST INDICATOR',
 'ABSENTEE OWNER STATUS',
 'PROPERTY LEVEL LATITUDE',
 'PROPERTY LEVEL LONGITUDE',
 'SITUS HOUSE NUMBER PREFIX',
 'SITUS HOUSE NUMBER',
 'SITUS HOUSE NUMBER SUFFIX',
 'SITUS DIRECTION',
 'SITUS STREET NAME',
 'SITUS MODE',
 'SITUS QUADRANT',
 'SITUS APARTMENT UNIT',
 'SITUS CITY',
 'SITUS STATE',
 'SITUS ZIP CODE',
 'SITUS CARRIER CODE',
 'MAILING HOUSE NUMBER PREFIX',
 'MAILING HOUSE NUMBER',
 'MAILING HOUSE NUMBER SUFFIX',
 'MAILING DIRECTION',
 'MAILING STREET NAME',
 'MAILING MODE',
 'MAILING QUADRANT',
 'MAILING APARTMENT UNIT',
 'MAILING PROPERTY CITY',
 'MAILING PROPERTY STATE',
 'MA

## Inspecting variable distributions

We can use the ```describe``` method to calculate summary statistics for select columns (variables). Summary statistics can only be calculated for columns of type ```integer``` or ```double```. Currently, all of our variables are stored as type ```string```. We can convert a column to numeric values by casting it to type ```integer```.

In [6]:
from pyspark.sql.types import IntegerType
df = df.withColumn("MORTGAGE AMOUNT", df["MORTGAGE AMOUNT"].cast(IntegerType()))
df.select(["MORTGAGE AMOUNT"]).dtypes

[('MORTGAGE AMOUNT', 'int')]

Now that the 'mortgage amount' column is numeric, we can calculate basic summary statistics (mean, standard deviation, minimum and maximum value). As you can see, the average mortgage amount in the CoreLogic data is about $800K.

In [7]:
df.describe(['MORTGAGE AMOUNT']).show()

+-------+--------------------+
|summary|     MORTGAGE AMOUNT|
+-------+--------------------+
|  count|           111058818|
|   mean|   779943.3151940263|
| stddev|1.6808331794028625E7|
|    min|                   0|
|    max|          2075000000|
+-------+--------------------+



Let's do the same for the sale amount of a property:

In [8]:
df = df.withColumn("SALE AMOUNT", df["SALE AMOUNT"].cast(IntegerType()))
df.describe(['SALE AMOUNT']).show()

+-------+-----------------+
|summary|      SALE AMOUNT|
+-------+-----------------+
|  count|        213239549|
|   mean|402407.0318453403|
| stddev|6648148.318833588|
|    min|                0|
|    max|       2128327250|
+-------+-----------------+



Interestingly, the average sale amount is about half of the average mortgage amount. But these differences could also be driven by missing information in the data.

## Exploring the amount of missing data within and across columns

It is important to know for columns of interest how many missing values they contain. Information on data missingness guides decisions on what needs to be done with affected rows. Can rows with missing information simply be dropped? Or, do we need to try to fill in missing values conditional on information in other columns?

Let's investigate whether the mortgage column contains any missing values.

In [8]:
from pyspark.sql.functions import isnan, when, count, col
df.select([count(when(isnan('MORTGAGE AMOUNT'),True))]).show()

+-----------------------------------------------------+
|count(CASE WHEN isnan(MORTGAGE AMOUNT) THEN true END)|
+-----------------------------------------------------+
|                                                    0|
+-----------------------------------------------------+



This column does not contain any missing values encoded as ```NaN```. Let's investigate whether the mortgage column containts any ```null``` values. This is just another way of how missing values could be stored.

In [10]:
df.select([count(when(col('MORTGAGE AMOUNT').isNull(),True))]).show()

+--------------------------------------------------------+
|count(CASE WHEN (MORTGAGE AMOUNT IS NULL) THEN true END)|
+--------------------------------------------------------+
|                                               256723662|
+--------------------------------------------------------+



That is quite a lot of missing information. We can calculate the percentage of ```null``` values in the mortgage column to get a better idea of the amount of missingness. To do so, we divide the number of rows with ```null``` values in the mortgage amount column by the total number of rows in the CoreLogic data. We can reuse the ```numRows``` variable here that we calculated earlier (see comment on 'lazy execution' above) to save us some processing time.

In [11]:
df.select([(count(when(col('MORTGAGE AMOUNT').isNull(),True))/numRows)]).show()

+----------------------------------------------------------------------+
|(count(CASE WHEN (MORTGAGE AMOUNT IS NULL) THEN true END) / 367782480)|
+----------------------------------------------------------------------+
|                                                    0.6980312439026459|
+----------------------------------------------------------------------+



As you can see, almost 70% of mortgage amount information is missing in the deeds dataset.

Let's calculate the proportion of records for which mortgage AND sale amount information is missing:

In [12]:
df.select([(count(when(col('MORTGAGE AMOUNT').isNull() & col('SALE AMOUNT').isNull(), True))/numRows)]).show()

+--------------------------------------------------------------------------------------------------+
|(count(CASE WHEN ((MORTGAGE AMOUNT IS NULL) AND (SALE AMOUNT IS NULL)) THEN true END) / 367782480)|
+--------------------------------------------------------------------------------------------------+
|                                                                               0.40302083720790616|
+--------------------------------------------------------------------------------------------------+



As we can see, about 40% of records have missing information on both the mortgage and sale amount columns.

## Inspecting categorical variables

We might be interested in the range of unique values of categorical variables. Let's print the total number of unique values of states designated by the 'SITUS STATE' column. (According to the Oxford Dictionary, "situs" is a legal term that designates "*the place to which, for purposes of legal jurisdiction or taxation, a property belongs.*")

In [20]:
nStates = len(df.select('SITUS STATE').distinct().collect())
print(nStates)

53

The SITUS STATE column contains 53 unique values, which is unexpected. There should be 50 states plus Washington, DC. 

Let's create a frequency table of the column values to get a better idea of this variable. We will sort the table in descending order of state counts. We will set ```n=nStates``` in the show() method to print all rows in the table. The two additional values are: (1) ```null``` (the amount of missing values in the SITUS STATE column), and (2) `VI`, which stands for the Virgin Islands.

In [12]:
from pyspark.sql.functions import desc
df.groupBy('SITUS STATE').count().sort(desc("count")).show(n=nStates)

+-----------+--------+
|SITUS STATE|   count|
+-----------+--------+
|         CA|52167596|
|       null|35371569|
|         FL|35281257|
|         TX|21674192|
|         OH|13347704|
|         PA|12254911|
|         AZ|12006118|
|         IL|11947950|
|         GA|11148302|
|         NY|10456027|
|         NC| 9686809|
|         MI| 8638900|
|         TN| 8638649|
|         CO| 8112459|
|         WA| 8104844|
|         VA| 7827533|
|         NJ| 7320302|
|         MD| 6525203|
|         MO| 6496410|
|         SC| 6311502|
|         OK| 5217718|
|         MA| 4876134|
|         OR| 4767128|
|         NV| 4712444|
|         IN| 4394358|
|         AL| 4379161|
|         AR| 4378459|
|         MN| 4267855|
|         WI| 3864221|
|         UT| 3154019|
|         KY| 3018343|
|         MS| 2978654|
|         HI| 2572320|
|         CT| 2513164|
|         LA| 2497956|
|         WV| 2374639|
|         KS| 2348984|
|         IA| 1982068|
|         NM| 1546001|
|         NE| 1365483|
|         M

## Inspecting string columns that contain date information

We might be interested in columns that present information about time/dates. Let's go ahead and identify the columns that seem to contain date information. As we can see below, dates are stored in   ```YYYYMMDD``` format in the deeds data and there is quite a bit of missing information (```null``` as well).

In [13]:
df.select([s for s in df.columns if 'DATE' in s]).show()

+--------------------+-------------------------+-------------+-----------------+
|SALE DATE (YYYYMMDD)|RECORDING DATE (YYYYMMDD)|MORTGAGE DATE|MORTGAGE DUE DATE|
+--------------------+-------------------------+-------------+-----------------+
|                null|                 19780600|         null|             null|
|                null|                 19910717|     19910717|             null|
|                null|                 20050908|         null|             null|
|                null|                 20050804|         null|             null|
|                null|                 20051229|         null|             null|
|                null|                 20051221|         null|             null|
|                null|                 20051221|         null|             null|
|                null|                 20051216|         null|             null|
|                null|                 20051214|         null|             null|
|                null|      

Let's work with the ```RECORDING DATE (YYYYMMDD)``` column to identify the time period (in years) that is covered in the CoreLogic data. We will rename this column first to get rid of the space and parentheses in the column name. 

In [14]:
df = df.withColumnRenamed('RECORDING DATE (YYYYMMDD)','RECORDINGDATE')

Now we are ready to subset this string and collect the first four digits to obtain year-only information. We will also cast this new column into type integer.

In [15]:
df = df.withColumn('record_year', df.RECORDINGDATE.substr(0,4).cast(IntegerType()))
df.select(['RECORDINGDATE','record_year']).show()

+-------------+-----------+
|RECORDINGDATE|record_year|
+-------------+-----------+
|     19780600|       1978|
|     19910717|       1991|
|     20050908|       2005|
|     20050804|       2005|
|     20051229|       2005|
|     20051221|       2005|
|     20051221|       2005|
|     20051216|       2005|
|     20051214|       2005|
|     20051205|       2005|
|     20051117|       2005|
|     20051109|       2005|
|     20051024|       2005|
|     20051031|       2005|
|     20051031|       2005|
|     20051028|       2005|
|     20051027|       2005|
|     20051020|       2005|
|     19910717|       1991|
|     20051014|       2005|
+-------------+-----------+
only showing top 20 rows



Now that we have created a new column that just gives the year of a deed record, let's investigate which years are represented in the CoreLogic data. Let's calculate the number of unique years first that are covered in the dataset.

In [18]:
nYears = len(df.select('record_year').distinct().collect())
print(nYears)

It looks like the CoreLogic data covers more than a century of deed records. We can calculate the frequency distribution of ```record_year``` by counting the number of observations for each unique ```record_year``` value. Let's also sort this distribution by ```record_year``` using the ```sort``` method. Using the ```show``` method with ```n``` equal to the number of unique record year values we calculated previously will display the frequency distribution in a table.

In [19]:
df.groupBy('record_year').count().sort("record_year").show(n=nYears)

+-----------+--------+
|record_year|   count|
+-----------+--------+
|       null|65650040|
|       1812|       1|
|       1896|       1|
|       1900|   14455|
|       1901|    5802|
|       1902|    3051|
|       1903|    2772|
|       1904|    2378|
|       1905|    2304|
|       1906|    2023|
|       1907|    2183|
|       1908|    1940|
|       1909|    2314|
|       1910|    5909|
|       1911|    4141|
|       1912|    3698|
|       1913|    3628|
|       1914|    3686|
|       1915|    2785|
|       1916|    3098|
|       1917|    2848|
|       1918|    2540|
|       1919|    3535|
|       1920|    3210|
|       1921|    2431|
|       1922|    2767|
|       1923|    3134|
|       1924|    3170|
|       1925|    3305|
|       1926|    3157|
|       1927|    3612|
|       1928|    3452|
|       1929|    3413|
|       1930|    3702|
|       1931|    2819|
|       1932|    1983|
|       1933|    2191|
|       1934|    2199|
|       1935|    2779|
|       1936|    3468|
|       193

If you are just interested in the minimum and maximum values of the ```record_year``` column, you can aggregate this column and apply the ```min```/```max``` functions to it:

In [22]:
minYear = df.agg({'record_year': 'min'}).collect()
maxYear = df.agg({'record_year': 'max'}).collect()
print(minYear, maxYear)

[Row(min(record_year)=1812)] [Row(max(record_year)=2017)]


## Filtering the CoreLogic data to create a subset of observations

Suppose we only needed a subset of the CoreLogic data for research purposes, i.e., we were interested in property information in a select location. 

Here, our goal is to filter the CoreLogic data to create a subset that only contains rows relating to Detroit, Michigan. 

To start, let's print a select few rows and columns of the dataframe:

In [15]:
df.select('FIPS', 'SITUS CITY', 'SITUS STATE', 'SITUS ZIP CODE').show()

+-----+----------+-----------+--------------+
| FIPS|SITUS CITY|SITUS STATE|SITUS ZIP CODE|
+-----+----------+-----------+--------------+
|12099|      null|       null|          null|
|17097|      null|       null|          null|
|97200|      null|       null|          null|
|97200|      null|       null|          null|
|97199|      null|       null|          null|
|97199|      null|       null|          null|
|97199|      null|       null|          null|
|97199|      null|       null|          null|
|97199|      null|       null|          null|
|97199|      null|       null|          null|
|97199|      null|       null|          null|
|97199|      null|       null|          null|
|97199|      null|       null|          null|
|97199|      null|       null|          null|
|97199|      null|       null|          null|
|97199|      null|       null|          null|
|97199|      null|       null|          null|
|97199|      null|       null|          null|
|17097|      null|       null|    

Notice how there are many rows containing columns with ```null``` values. We will remove rows that have ```null``` values across all columns of interest because it would be hard to deduce the location without any of this information (FIPS code, property city, state and zip code) present.

In [16]:
df = df.dropna(subset=["FIPS","SITUS CITY", "SITUS STATE", "SITUS ZIP CODE"])

Now let's see how many rows we have left out of the original 367,782,480:

In [17]:
df.count()

311359183

Let's assume we wanted to filter the CoreLogic data for Wayne county, MI, where the city of Detroit is located. The FIPS code for Wayne county is '26163' ([source](https://mi.postcodebase.com/county/26163)).

In [18]:
WayneCsubset = df.filter(df.FIPS==26163)

Let's print a few rows and columns of the subset. As you can see, we are only dealing with Wayne county now.

In [19]:
WayneCsubset.select('FIPS', 'SITUS CITY', 'SITUS STATE', 'SITUS ZIP CODE').show()

+-----+----------+-----------+--------------+
| FIPS|SITUS CITY|SITUS STATE|SITUS ZIP CODE|
+-----+----------+-----------+--------------+
|26163|   DETROIT|         MI|     482013148|
|26163|   DETROIT|         MI|     482013148|
|26163|   DETROIT|         MI|     482013148|
|26163|   DETROIT|         MI|     482013148|
|26163|   DETROIT|         MI|     482012463|
|26163|   DETROIT|         MI|         48202|
|26163|   DETROIT|         MI|         48202|
|26163|   DETROIT|         MI|         48202|
|26163|   DETROIT|         MI|     482022828|
|26163|   DETROIT|         MI|     482022828|
|26163|   DETROIT|         MI|     482022828|
|26163|   DETROIT|         MI|     482022828|
|26163|   DETROIT|         MI|     482022828|
|26163|   DETROIT|         MI|         48202|
|26163|   DETROIT|         MI|     482021302|
|26163|   DETROIT|         MI|     482021368|
|26163|   DETROIT|         MI|     482021368|
|26163|   DETROIT|         MI|     482021368|
|26163|   DETROIT|         MI|    

How many rows are in the Wayne county subset? We can use the count() method:

In [20]:
WayneCsubset.count()

2132959

There are many different cities within Wayne county. Let's get a list of all of them:

In [21]:
WayneCsubset.select('SITUS CITY').distinct().collect()

[Row(SITUS CITY='LINCOLN PARK'),
 Row(SITUS CITY='MELVINDALE'),
 Row(SITUS CITY='NORTHVILLE'),
 Row(SITUS CITY='TRENTON'),
 Row(SITUS CITY='YPSILANTI'),
 Row(SITUS CITY='BELLEVILLE'),
 Row(SITUS CITY='OAK PARK'),
 Row(SITUS CITY='RIVERVIEW'),
 Row(SITUS CITY='GROSSE POINTE WOODS'),
 Row(SITUS CITY='FLAT ROCK'),
 Row(SITUS CITY='GROSSE ILE'),
 Row(SITUS CITY='GROSSE POINTE PARK'),
 Row(SITUS CITY='WARREN'),
 Row(SITUS CITY='GIBRALTAR'),
 Row(SITUS CITY='GROSSE POINTE'),
 Row(SITUS CITY='MONROE'),
 Row(SITUS CITY='WYANDOTTE'),
 Row(SITUS CITY='HIGHLAND'),
 Row(SITUS CITY='WOODHAVEN'),
 Row(SITUS CITY='INKSTER'),
 Row(SITUS CITY='SOUTHGATE'),
 Row(SITUS CITY='SUMPTER TWP'),
 Row(SITUS CITY='GARDEN CITY'),
 Row(SITUS CITY='HIGHLAND PARK'),
 Row(SITUS CITY='DETROIT'),
 Row(SITUS CITY='FERNDALE'),
 Row(SITUS CITY='CHELSEA'),
 Row(SITUS CITY='BROWNSTOWN TWP'),
 Row(SITUS CITY='NEW HUDSON'),
 Row(SITUS CITY='LANSING'),
 Row(SITUS CITY='DEARBORN HEIGHTS'),
 Row(SITUS CITY='CANTON'),
 Row(SITUS 

Let's create a subset for Detroit only. We will filter using the 'SITUS CITY' variable. This variables designates the city associated with the property address. 

Note that this is a straightforward filtering method if using the Wayne county subset we created previously. 

In [22]:
detroit = WayneCsubset.filter(WayneCsubset["SITUS CITY"]=="DETROIT")

If you wanted to create a Detroit subset from the entire dataset, we suggest filtering on 'SITUS CITY' and 'SITUS STATE'. When using the entire dataframe, filtering on 'SITUS CITY' alone could run into city name ambiguity across states. 

In [23]:
detroit2 = df.filter((df["SITUS CITY"]=="DETROIT") & (df["SITUS STATE"]=="MI"))

How many rows are in the Detroit subset? Depending on the size of the dataframe, count() can take a bit of time to calculate. Therefore, we will save the value to a variable in case we need to use it later on. This can save us some processing time (cf. comment on "lazy evaluation" above).

In [24]:
detroitCount = detroit.count()
detroitCount

1129407

Do the two filtering methods yield the same number of rows?

In [25]:
detroit2Count = detroit2.count()
detroit2Count

1129438

In [26]:
detroit2Count - detroitCount

31

The second subset contains slightly more records. This means that we get two different results depending on whether we filter by (SITUS CITY == "DETROIT" AND SITUS STATE == "MI") versus (FIPS == "26163" AND SITUS CITY == "DETROIT"). 

Let's investigate the Detroit subset with more records (```detroit2```). We can tabulate the FIPS codes that have city=="DETROIT" and state=="MI": 

In [27]:
detroit2.groupBy('FIPS', 'SITUS CITY', 'SITUS STATE').count().show()

+-----+----------+-----------+-------+
| FIPS|SITUS CITY|SITUS STATE|  count|
+-----+----------+-----------+-------+
|26147|   DETROIT|         MI|      6|
|26061|   DETROIT|         MI|      2|
|26035|   DETROIT|         MI|      2|
|26161|   DETROIT|         MI|      1|
|26163|   DETROIT|         MI|1129407|
|48039|   DETROIT|         MI|      2|
|26069|   DETROIT|         MI|      2|
|26071|   DETROIT|         MI|      5|
|26099|   DETROIT|         MI|      2|
|26115|   DETROIT|         MI|      9|
+-----+----------+-----------+-------+



As you can see, there are multiple FIPS codes associated with Detroit, MI, in this data subset when there should really only be one FIPS code for Detroit in Wayne County, Michigan.  For example, FIPS code '26115' (9 occurences) is for Monroe County which does not contain the city of Detroit. 

For each of the additional FIPS codes identified above, we would have to do some further research to determine whether these FIPS codes constitute data entry errors that need to be corrected or whether they identify properties that are not actually located in Detroit. To do this, we would have to review each FIPS code and also inspect the addresses associated with each property to determine where exactly a property is legally located. This work is beyond the purposes of the current analysis.

## Saving a subset of the CoreLogic data to a spreadsheet

Let's assume the detroit2 dataframe looks like what we are interested in and we now want to write this Detroit subset to a new csv file in an ```output/``` folder in our ```home/[uniqname]/``` directory. (Note: You probably need to change this path, depending on where you store this notebook.) 

By default, Spark writes big data into multiple files to optimize computation times during import and export. This is recommended practice, especially if the data is big, i.e., if it has many rows and/or columns.

The below command writes 100 partitions of the Detroit data (i.e., 100 csv files) to the designated folder. 

_Note_: If you saved data to this path before, you will need to add the argument ```overwrite=True``` to the function call like so ```detroit2.write.csv("../../output/corelogic_data_deeds_Detroit_partitioned", overwrite=True)```

In [28]:
detroit2.write.csv("../../output/corelogic_data_deeds_Detroit_partitioned")

There is also a way to force the data into one single file. Sometimes, a single file is needed to continue processing the data with other research software. Saving the data into one file is a little risky and might not always work, so caution is advised. 

The following command combines the data into one core (Caution: it is very slow). If the data does not fit, it will throw an error. _Note_: If you saved data to this path before, you will need to add the argument ```overwrite=True``` to the function call. The overwrite argument would overwrite the previously created ```output/``` directory.

Note: You would only need to run this, if you indeed need only one single data file. Many research programs allow you to read in and combine multiple data files into one object. See for example the [glob method in Python](https://docs.python.org/3/library/glob.html).

In [29]:
detroit2.coalesce(1).write.csv("../../output/corelogic_data_deeds_Detroit_singlefile")

## Drawing a data sample for research and development purposes
Let's say we want to draw and save just a sample of the Detroit data (n~1000) that is small enough to use for writing and testing code for analyzing the Detroit subset. We can use PySpark's DataFrame.sample() method. However, we have to specify a fraction of the dataframe size rather than explicitly the size of the sample, n. Also, we are not guaranteed to get EXACTLY the fraction that we specify, due to the way the sample() function is implemented. Let's see how that works.

Since we want about 1000 rows in our sample, we can divide 1000 by detroit2Count to see what fraction that is.

In [30]:
sampleFraction = 1000 / detroit2Count
sampleFraction

0.0008853960996531018

Now, let's sample the Detroit data.

In [31]:
detroitSample = detroit2.sample(fraction=sampleFraction)

How many records do we have in our sample?

In [32]:
detroitSample.count()

1002

Now that we have created a small sample of our Detroit records for research and development purposes, we can write the results to a spreadsheet (note that using coalesce(1) can still be quite expensive. Use only if absolutely necessary).

In [33]:
detroitSample.write.csv("../../output/corelogic_data_deeds_Detroit_sampled_1031_singlefile")